module/module.go

```diff
@@ -0,0 +1,501 @@
+				classInstance.ClassName,
+				classInstance.InstanceName,
+				buildPath,
+				moduleIndex,
+				moduleCount,
+			)
+
+			classGenerators := moduleSystem.classes[classInstance.ClassName]
+			generator := classGenerators.types[classInstance.ClassType]
+
+			if generator == nil {
+				fmt.Printf(
+					"Skipping generation of %s %s class of type %s as generator is not defined\n",
+					classInstance.InstanceName,
+					classInstance.ClassName,
+					classInstance.ClassType,
+				)
+				continue
+			}
+
+			files, err := generator.Generate(classInstance)
+
+			if err != nil {
+				fmt.Printf(
+					"Error generating %s %s class %s type as generator is not defined\n%s\n",
+					classInstance.InstanceName,
+					classInstance.ClassName,
+					classInstance.ClassType,
+					err.Error(),
+				)
+				return err
+			}
+
+			for filePath, content := range files {
+				filePath = filepath.Clean(filePath)
+
+				if strings.HasPrefix(filePath, "..") {
+					return errors.Errorf(
+						"Module %s generated a file outside the build dir %s",
+						classInstance.Directory,
+						filePath,
+					)
+				}
+
+				resolvedPath := filepath.Join(
+					buildPath,
+					filePath,
+				)
+
+				if err := ioutil.WriteFile(
+					resolvedPath,
+					content,
+					0644,
+				); err != nil {
+					return err
+				}
+
+				// HACK: The module system writer shouldn't
+				// assume that we want to format the files in
+				// this way, but we don't have these formatters
+				// as a library or a custom post build script
+				// for the generators yet.
+				if filepath.Ext(filePath) == ".go" {
+					if err := formatGoFile(resolvedPath); err != nil {
+						return err
+					}
+				}
+			}
+		}
+	}
+
+	print("\nDone.\n")
+	return nil
+}
+
+func formatGoFile(filePath string) error {
+	gofmtCmd := exec.Command("gofmt", "-s", "-w", "-e", filePath)
+	gofmtCmd.Stdout = os.Stdout
+	gofmtCmd.Stderr = os.Stderr
+
+	if err := gofmtCmd.Run(); err != nil {
+		return errors.Wrapf(err, "failed to gofmt file: %s", filePath)
+	}
+
+	goimportsCmd := exec.Command("goimports", "-w", "-e", filePath)
+	goimportsCmd.Stdout = os.Stdout
+	goimportsCmd.Stderr = os.Stderr
+
+	if err := goimportsCmd.Run(); err != nil {
+		return errors.Wrapf(err, "failed to goimports file: %s", filePath)
+	}
+
+	return nil
+}
+
+// Class defines a module class in the build configuration directory. This
+// coud be something like an Endpoint class which contains multiple endpoint
+// configurations, or a Lib class, that is itself a module instance
+type Class struct {
+	ClassType         classType
+	Directory         string
+	ClassDependencies []string
+	types             map[string]BuildGenerator
+}
+
+// BuildGenerator provides a function to generate a module instance build
+// artifact from its configuration as part of a build step. For example, an
+// Endpoint module instance may generate endpoint handler code
+type BuildGenerator interface {
+	Generate(
+		instance *Instance,
+	) (map[string][]byte, error)
```

Comment:
It's also wrong to say the Instance has all the information required. I'm actually building to a hard coded directory composed from the BaseDirectory, the Directory and the "build" directory.. This is a slight regression compared with how we currently do it, it's a TBD.

I would like to remove BaseDirectory from the instance as it shouldn't be required, but it is necessary to produce the clients at this point I believe.